### PR TITLE
feat: add trestle beta command

### LINKED ISSUE #2258: trestle beta command
Allow introduction of new commands that are in beta test mode.

Proposed for discussion:

# Trestle Beta Command Specification

## Overview
The `trestle beta` command provides a mechanism to manage experimental/beta features in compliance-trestle. This allows developers to ship new functionality that can be selectively enabled by users, providing a safe way to test and iterate on features before they become stable.

## Design Goals
1. **Safe experimentation**: Allow users to opt-in to beta features
2. **Clear visibility**: Users can see what beta features are available
3. **Easy management**: Simple commands to enable/disable features
4. **Self-documenting**: Beta commands identify themselves and check if enabled
5. **Persistent configuration**: Beta feature state persists across sessions

## Command Structure

```
trestle beta <subcommand> [options]
```

#### 1. `help`
Display help information about the beta command system.

```bash
trestle beta help
```

**Output:**
```
Trestle Beta Feature Management

The beta command allows you to manage experimental features in trestle.
Beta features are under active development and may change or be removed
in future releases.

Usage:
  trestle beta help              Show this help message
  trestle beta query             List all beta features and their status
  trestle beta enable <feature>  Enable a beta feature
  trestle beta disable <feature> Disable a beta feature

Examples:
  trestle beta query
  trestle beta enable sign
  trestle beta disable verify

For more information about specific beta features, use:
  trestle beta query --verbose
```

#### 2. `query`
List all available beta features and their current status.

```bash
trestle beta query [--verbose]
```

**Output (default):**
```
Beta Features Status:

Enabled:
  ✓ sign               Cryptographic signing of OSCAL documents

Available (Disabled):
  ✗ verify             Cryptographic verification of signed OSCAL documents
  ✗ ai-assist          AI-powered control mapping assistance

Use 'trestle beta query --verbose' for detailed descriptions.
Use 'trestle beta enable <feature>' to enable a feature.
```

**Output (verbose):**
```
Beta Features Status:

Enabled:
  ✓ sign
    Description: Cryptographic signing of OSCAL documents
    Status: Enabled
    Since: v4.1.0
    Stability: Beta
    Commands: trestle sign
    Documentation: https://oscal-compass.github.io/compliance-trestle/beta/sign

Available (Disabled):
  ✗ verify
    Description: Cryptographic verification of signed OSCAL documents
    Status: Disabled
    Since: v4.1.0
    Stability: Beta
    Commands: trestle verify
    Documentation: https://oscal-compass.github.io/compliance-trestle/beta/verify
    
  ✗ ai-assist
    Description: AI-powered control mapping assistance using LLMs
    Status: Disabled
    Since: v4.3.0
    Stability: Experimental
    Commands: trestle author controls --ai-assist
    Documentation: https://oscal-compass.github.io/compliance-trestle/beta/ai-assist
```

#### 3. `enable`
Enable a specific beta feature.

```bash
trestle beta enable <feature-name>
```

**Examples:**
```bash
trestle beta enable sign
trestle beta enable verify
```

**Output (success):**
```
✓ Beta feature 'sign' has been enabled.

This feature is experimental and may change in future releases.
Documentation: https://oscal-compass.github.io/compliance-trestle/beta/sign

Available commands:
  trestle sign --help

To disable this feature, run:
  trestle beta disable sign
```

**Output (already enabled):**
```
ℹ Beta feature 'sign' is already enabled.
```

**Output (feature not found):**
```
✗ Error: Beta feature 'unknown-feature' not found.

Available beta features:
  - sign
  - verify
  - ai-assist

Use 'trestle beta query' to see all available features.
```

#### 4. `disable`
Disable a specific beta feature.

```bash
trestle beta disable <feature-name>
```

**Examples:**
```bash
trestle beta disable sign
trestle beta disable verify
```

**Output (success):**
```
✓ Beta feature 'sign' has been disabled.

The following commands will no longer be available:
  trestle sign

To re-enable this feature, run:
  trestle beta enable sign
```

**Output (already disabled):**
```
ℹ Beta feature 'sign' is already disabled.
```

**Output (feature not found):**
```
✗ Error: Beta feature 'unknown-feature' not found.

Use 'trestle beta query' to see all available features.
```

## Configuration Storage

Beta feature configuration is stored in the trestle workspace configuration:

**Location:** `.trestle/config.ini`

**Format:**
```ini
[beta]
enabled_features = sign,verify
```

**Fallback:** If no workspace is found, configuration is stored in user home directory:
- Linux/Mac: `~/.config/trestle/beta.ini`
- Windows: `%APPDATA%\trestle\beta.ini`

### Feature Definition
Each beta feature is defined with the following attributes:

```python
@dataclass
class BetaFeature:
    """Definition of a beta feature."""
    
    name: str                    # Unique identifier (kebab-case)
    description: str             # Short description
    commands: List[str]          # Commands provided by this feature
    since_version: str           # Version when introduced
    stability: str               # 'experimental', 'alpha', 'beta'
    documentation_url: str       # Link to documentation
    enabled_by_default: bool     # Whether enabled by default
    deprecation_version: Optional[str]  # Version when deprecated
```

### Registration Example

```python
# In trestle/core/beta_features.py

BETA_FEATURES = {
    'sign': BetaFeature(
        name='sign',
        description='Cryptographic signing of OSCAL documents',
        commands=['trestle sign'],
        since_version='4.1.0',
        stability='beta',
        documentation_url='https://oscal-compass.github.io/compliance-trestle/beta/sign',
        enabled_by_default=False,
        deprecation_version=None
    ),
    'verify': BetaFeature(
        name='verify',
        description='Cryptographic verification of signed OSCAL documents',
        commands=['trestle verify'],
        since_version='4.1.0',
        stability='beta',
        documentation_url='https://oscal-compass.github.io/compliance-trestle/beta/verify',
        enabled_by_default=False,
        deprecation_version=None
    ),
}
```

### Decorator for Beta Commands

Commands can declare themselves as beta features using a decorator:

```python
from trestle.core.beta_features import beta_feature, is_beta_enabled

class SignCommand(CommandBase):
    """Sign OSCAL documents with cryptographic signatures."""
    
    name = 'sign'
    
    @beta_feature('sign')
    def _run(self, args: argparse.Namespace) -> int:
        """Run the sign command."""
        # Command implementation
        logger.info('Signing OSCAL document...')
        # ... signing logic ...
        return CmdReturnCodes.SUCCESS.value
```

### Beta Feature Check

The decorator automatically checks if the feature is enabled:

```python
def beta_feature(feature_name: str):
    """Decorator to mark a command as requiring a beta feature."""
    def decorator(func):
        @functools.wraps(func)
        def wrapper(self, *args, **kwargs):
            if not is_beta_enabled(feature_name):
                feature = BETA_FEATURES.get(feature_name)
                print(f"✗ Error: The '{self.name}' command is a beta feature and is not enabled.")
                print(f"\nFeature: {feature_name}")
                print(f"Description: {feature.description}")
                print(f"Stability: {feature.stability}")
                print(f"Documentation: {feature.documentation_url}")
                print(f"\nTo enable this feature, run:")
                print(f"  trestle beta enable {feature_name}")
                return 1
            return func(self, *args, **kwargs)
        return wrapper
    return decorator
```

### Command Registration

Beta commands should be registered conditionally based on whether the feature is enabled:

```python
# In trestle/cli.py

def setup_commands(parser: argparse.ArgumentParser) -> None:
    """Setup all available commands."""
    subparsers = parser.add_subparsers(dest='command')
    
    # Standard commands
    setup_command(subparsers, ValidateCommand)
    setup_command(subparsers, ImportCommand)
    # ... other standard commands ...
    
    # Beta commands - only register if enabled
    if is_beta_enabled('sign'):
        setup_command(subparsers, SignCommand)
    
    if is_beta_enabled('verify'):
        setup_command(subparsers, VerifyCommand)
```

### Scenario 1: User tries to use a beta command without enabling it

```bash
$ trestle sign catalog.json --key private.pem

✗ Error: The 'sign' command is a beta feature and is not enabled.

Feature: sign
Description: Cryptographic signing of OSCAL documents
Stability: beta
Documentation: https://oscal-compass.github.io/compliance-trestle/beta/sign

To enable this feature, run:
  trestle beta enable sign
```

### Scenario 2: User enables a beta feature and uses it

```bash
$ trestle beta enable sign

✓ Beta feature 'sign' has been enabled.

This feature is experimental and may change in future releases.
Documentation: https://oscal-compass.github.io/compliance-trestle/beta/sign

Available commands:
  trestle sign --help

To disable this feature, run:
  trestle beta disable sign

$ trestle sign catalog.json --key private.pem
Signing OSCAL document catalog.json...
✓ Document signed successfully
Signature written to catalog.json.sig
```

### Scenario 3: User queries available features

```bash
$ trestle beta query

Beta Features Status:

Enabled:
  ✓ sign               Cryptographic signing of OSCAL documents

Available (Disabled):
  ✗ verify             Cryptographic verification of signed OSCAL documents
  ✗ ai-assist          AI-powered control mapping assistance

Use 'trestle beta query --verbose' for detailed descriptions.
```

### Scenario 4: User enables both sign and verify

```bash
$ trestle beta enable verify

✓ Beta feature 'verify' has been enabled.

This feature is experimental and may change in future releases.
Documentation: https://oscal-compass.github.io/compliance-trestle/beta/verify

Available commands:
  trestle verify --help

To disable this feature, run:
  trestle beta disable verify

$ trestle verify catalog.json --signature catalog.json.sig --key public.pem
Verifying signature for catalog.json...
✓ Signature is valid
Signed by: CN=OSCAL Authority, O=Example Org
Signed at: 2026-06-12T16:45:00Z
```

### Scenario 5: Help shows beta commands only when enabled

```bash
$ trestle --help
usage: trestle [-h] [--version] [--verbose] {validate,import,author,...} ...

# sign and verify not shown

$ trestle beta enable sign

$ trestle --help
usage: trestle [-h] [--version] [--verbose] {validate,import,author,...,sign} ...

Available commands:
  ...
  sign                Sign OSCAL documents (BETA)
  ...
```

## Implementation Checklist

- [ ] Create `trestle/core/beta_features.py` module
  - [ ] Define `BetaFeature` dataclass
  - [ ] Implement `BETA_FEATURES` registry
  - [ ] Implement `is_beta_enabled()` function
  - [ ] Implement `get_enabled_features()` function
  - [ ] Implement `enable_feature()` function
  - [ ] Implement `disable_feature()` function
  - [ ] Implement `@beta_feature` decorator

- [ ] Create `trestle/commands/beta.py` command
  - [ ] Implement `help` subcommand
  - [ ] Implement `query` subcommand
  - [ ] Implement `enable` subcommand
  - [ ] Implement `disable` subcommand

- [ ] Configuration management
  - [ ] Implement config file reading/writing
  - [ ] Handle workspace vs. user-level config
  - [ ] Add migration for existing configs

- [ ] Update CLI registration
  - [ ] Modify `trestle/cli.py` to conditionally register beta commands
  - [ ] Add beta indicator in help text

- [ ] Create example beta commands
  - [ ] Implement `trestle sign` command
  - [ ] Implement `trestle verify` command

- [ ] Documentation
  - [ ] Add beta command documentation
  - [ ] Create beta feature documentation template
  - [ ] Document how to create beta features
  - [ ] Add sign/verify command documentation

- [ ] Testing
  - [ ] Unit tests for beta feature management
  - [ ] Integration tests for beta commands
  - [ ] Test decorator functionality
  - [ ] Test configuration persistence
  - [ ] Test sign/verify commands

## Example: Sign Command Implementation

```python
# trestle/commands/sign.py

class SignCommand(CommandBase):
    """Sign OSCAL documents with cryptographic signatures."""
    
    name = 'sign'
    
    def _init_arguments(self) -> None:
        self.add_argument('file', help='OSCAL file to sign', type=str)
        self.add_argument('-k', '--key', help='Private key file', required=True, type=str)
        self.add_argument('-o', '--output', help='Output signature file', type=str)
        self.add_argument('-a', '--algorithm', help='Signature algorithm', 
                         default='RS256', choices=['RS256', 'RS384', 'RS512'])
    
    @beta_feature('sign')
    def _run(self, args: argparse.Namespace) -> int:
        """Run the sign command."""
        try:
            logger.info(f'Signing OSCAL document {args.file}...')
            
            # Load the OSCAL document
            oscal_data = pathlib.Path(args.file).read_text()
            
            # Load private key
            with open(args.key, 'rb') as key_file:
                private_key = load_pem_private_key(key_file.read(), password=None)
            
            # Create signature
            signature = private_key.sign(
                oscal_data.encode(),
                padding.PSS(
                    mgf=padding.MGF1(hashes.SHA256()),
                    salt_length=padding.PSS.MAX_LENGTH
                ),
                hashes.SHA256()
            )
            
            # Write signature
            output_file = args.output or f'{args.file}.sig'
            pathlib.Path(output_file).write_bytes(signature)
            
            logger.info(f'✓ Document signed successfully')
            logger.info(f'Signature written to {output_file}')
            
            return CmdReturnCodes.SUCCESS.value
            
        except Exception as e:
            logger.error(f'Failed to sign document: {e}')
            return CmdReturnCodes.COMMAND_ERROR.value
```

## Example: Verify Command Implementation

```python
# trestle/commands/verify.py

class VerifyCommand(CommandBase):
    """Verify cryptographic signatures on OSCAL documents."""
    
    name = 'verify'
    
    def _init_arguments(self) -> None:
        self.add_argument('file', help='OSCAL file to verify', type=str)
        self.add_argument('-s', '--signature', help='Signature file', required=True, type=str)
        self.add_argument('-k', '--key', help='Public key file', required=True, type=str)
    
    @beta_feature('verify')
    def _run(self, args: argparse.Namespace) -> int:
        """Run the verify command."""
        try:
            logger.info(f'Verifying signature for {args.file}...')
            
            # Load the OSCAL document
            oscal_data = pathlib.Path(args.file).read_text()
            
            # Load signature
            signature = pathlib.Path(args.signature).read_bytes()
            
            # Load public key
            with open(args.key, 'rb') as key_file:
                public_key = load_pem_public_key(key_file.read())
            
            # Verify signature
            public_key.verify(
                signature,
                oscal_data.encode(),
                padding.PSS(
                    mgf=padding.MGF1(hashes.SHA256()),
                    salt_length=padding.PSS.MAX_LENGTH
                ),
                hashes.SHA256()
            )
            
            logger.info('✓ Signature is valid')
            
            return CmdReturnCodes.SUCCESS.value
            
        except InvalidSignature:
            logger.error('✗ Signature verification failed: Invalid signature')
            return CmdReturnCodes.COMMAND_ERROR.value
        except Exception as e:
            logger.error(f'Failed to verify signature: {e}')
            return CmdReturnCodes.COMMAND_ERROR.value
```

## Future Enhancements

1. **Feature Dependencies**: Allow features to depend on other features
   - Example: `verify` could depend on `sign`
2. **Feature Expiration**: Automatically disable features after a certain version
3. **Telemetry**: Optional anonymous usage tracking for beta features
4. **Feature Graduation**: Automated process to promote beta → stable
5. **Environment Variables**: Support `TRESTLE_BETA_FEATURES=sign,verify`
6. **CI/CD Integration**: Special handling for CI environments
7. **Feature Warnings**: Show warnings when using beta features in production

## Notes

- Beta features should have comprehensive documentation
- Breaking changes in beta features are acceptable
- Beta features should have clear graduation criteria
- Consider adding warnings when using beta features in production
- Beta feature state should be included in `trestle version` output
- Commands should gracefully handle when beta features are disabled mid-session

### PR DESCRIPTION
**Pull request by MatteoFari**
## Types of changes

- [ ] Hot fix (emergency fix and release)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [x] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Summary

Adds `trestle beta` for managing experimental features. Beta features can be queried, enabled, or disabled through the CLI, persisted in workspace/user config, enabled temporarily through TRESTLE_BETA_FEATURES, and guarded in code with the new beta feature decorator.
## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)
- fixes https://github.com/oscal-compass/compliance-trestle/issues/2258

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

File: docs/tutorials/cli.md [hunk 1]
```diff
@@ -179,6 +181,29 @@ Running `trestle version -n nist -t catalog` will return:
...
+
+Use `trestle beta enable <feature>` and `trestle beta disable <feature>` to manage a feature.
+
+```bash
+trestle beta enable example-feature
+trestle beta disable example-feature
+```
+
+When the current directory is inside a trestle workspace, beta feature state is stored in `.trestle/config.ini`.
+Outside a workspace, trestle uses a user-level beta config file.
```
Comment (degenaro) [inline docs/tutorials/cli.md]: Hmmm....where is this outside file and what is ti called?

File: tests/trestle/core/commands/beta_test.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
+# -*- mode:python; coding:utf-8 -*-
+
+# Copyright (c) 2026 IBM Corp. All rights reserved.
```
Comment (degenaro) [inline tests/trestle/core/commands/beta_test.py]: Please use oscal compass authors heading.

File: tests/trestle/core/beta_features_test.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
+# -*- mode:python; coding:utf-8 -*-
+
+# Copyright (c) 2026 IBM Corp. All rights reserved.
```
Comment (degenaro) [inline tests/trestle/core/beta_features_test.py]: Please use oscal compass authors heading.

File: trestle/core/commands/beta.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
+# -*- mode:python; coding:utf-8 -*-
+
+# Copyright (c) 2026 IBM Corp. All rights reserved.
```
Comment (degenaro) [inline trestle/core/commands/beta.py]: Please use oscal compass authors heading.

File: trestle/core/beta_features.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
+# -*- mode:python; coding:utf-8 -*-
+
+# Copyright (c) 2026 IBM Corp. All rights reserved.
```
Comment (degenaro) [inline trestle/core/beta_features.py]: Please use oscal compass authors heading.

File: docs/tutorials/cli.md (hunk 1 above)
Comment (MatteoFari) [reply docs/tutorials/cli.md]: I made this explicit in `docs/tutorials/cli.md`: outside a workspace, trestle uses `$XDG_CONFIG_HOME/trestle/beta.ini`, or `~/.config/trestle/beta.ini` if `XDG_CONFIG_HOME` is unset. On Windows it uses `%APPDATA%\trestle\beta.ini`.

**Comment by degenaro** [comment]
**Bob's review:**

Code Review: PR #2267 - Add trestle beta command
Overview
This PR introduces a comprehensive beta feature management system for compliance-trestle, allowing experimental features to be opt-in and properly managed. The implementation is well-designed, thoroughly tested, and properly documented.

Summary of Changes
New command: trestle beta for managing experimental features
Core infrastructure: Beta feature registration, configuration, and decorator system
CLI integration: Added BetaCmd to main CLI and beta status to version command
Documentation: Developer guide and user-facing CLI documentation
Tests: Comprehensive test coverage (375 new lines of tests)
Strengths
1. Excellent Architecture
Clean separation of concerns between command (beta.py) and core logic (beta_features.py)
Well-designed BetaFeature dataclass with all necessary metadata
Decorator pattern (@beta_feature) elegantly guards beta commands
Multiple configuration sources (workspace config, user config, environment variable)
2. Robust Configuration Management
Workspace-level config (.trestle/config.ini) for team settings
User-level fallback config for non-workspace usage
Environment variable (TRESTLE_BETA_FEATURES) for CI/CD scenarios
Proper config preservation when updating beta settings
Cross-platform support (Windows/Unix config paths)
3. Excellent User Experience
Clear, helpful error messages with actionable guidance
Verbose mode for detailed feature information
Shows available commands when enabling features
Prevents disabling environment-enabled or default-enabled features with clear error messages
Integration with trestle version to show enabled features
4. Comprehensive Testing
172 lines in beta_test.py covering all command scenarios
203 lines in beta_features_test.py covering core functionality
Tests for edge cases (already enabled/disabled, unknown features, environment conflicts)
Tests for decorator behavior
Tests for config preservation and cross-platform paths
5. Good Documentation
Developer guide (beta_features.md) with clear examples
Updated CLI tutorial mentioning beta features
Inline code documentation and docstrings
Documentation checklist for future beta features
Minor Observations
1. Type Hints
The code uses modern Python type hints consistently, including:

Set[str] and Dict[str, BetaFeature]
Optional[str] for nullable fields
Callable and TypeVar for decorator typing
Union types with | syntax (Python 3.10+)
This is excellent and maintains consistency with the codebase.

2. Error Handling
Proper exception handling with TrestleError and TrestleIncorrectArgsError
Generic exception handler in _run methods
Clear error messages that guide users to solutions
3. Code Quality
Follows project conventions (copyright headers, encoding declarations)
Consistent naming conventions
Proper use of pathlib for cross-platform path handling
No obvious security issues
4. Potential Enhancements (not blockers)
Consider adding a --json output format for trestle beta query for programmatic use
The enabled_by_default feature could use an example in the codebase
Consider adding telemetry/logging when beta features are used (for deprecation planning)
Test Coverage Analysis
The tests cover:

✅ All command actions (help, query, enable, disable)
✅ Verbose and non-verbose output
✅ Feature already enabled/disabled scenarios
✅ Unknown feature handling
✅ Environment variable configuration
✅ Workspace vs user config paths
✅ Config preservation
✅ Decorator blocking and allowing behavior
✅ Default-enabled features
✅ Deprecation version display
Documentation Quality
Clear developer guide with code examples
Proper markdown formatting
Explains configuration hierarchy
Includes checklist for future features
Updated CLI overview appropriately
Recommendations
✅ Approve with Confidence
This PR is production-ready. The implementation is:

Well-architected and maintainable
Thoroughly tested with excellent coverage
Properly documented for both users and developers
Follows project conventions and best practices
Provides clear user experience
Suggestions for Future Iterations (Optional)
Add JSON output format for programmatic queries
Consider adding usage analytics for beta features
Add example of enabled_by_default feature in codebase
Consider adding trestle beta list as alias for query
Conclusion
This is a high-quality PR that adds valuable functionality to compliance-trestle. The beta feature system will enable safer experimentation with new features while maintaining stability for production users. The code quality, testing, and documentation all meet professional standards.

Recommendation: APPROVE ✅

**Review by degenaro — APPROVED** [review]
LGTM

**Comment by degenaro** [comment]
@MatteoFari Is beta command ready for review? Do you intend to make any further changes? Else once @butler54  or @vikas-agarwal76 approve I'll push to develop. LGTM.

**Comment by MatteoFari** [comment]
@degenaro yes I didn't plan on making more changes, so it's ready for review, let me know if I should change/add something

**Comment by MatteoFari** [comment]
@degenaro The tests create a dummy beta command in `tests/trestle/core/beta_features_test.py` at line 187, then wrap it with `@beta_feature('sample')` at line 200.

That covers both cases:
- when disabled, the dummy command is blocked and tells the user how to enable it
- when enabled, the dummy command runs normally like any other command

If a feature is not registered in `BETA_FEATURES`, `trestle beta enable <name>` just returns `Beta feature '<name>' not found.`

**Comment by MatteoFari** [comment]
Adds `--beta` as a one-time opt-in for beta-gated commands without persisting config. The flag is hidden from normal command help, rejected by non-beta commands, and shown only in help for commands with a beta feature decorator.

**Comment by MatteoFari** [comment]
@degenaro LGTM
